### PR TITLE
Composer repository

### PR DESCRIPTION
**Pull request by Nicofuma**
Fix the old implementation in order to make titania acting as a valid composer repository.

A few important things:
- Any package in the 'composer compatible types' list and with a composer.json file is added to the index
- Only validated packages should be present
- We use the original composer.json file instead of building a new one. This way we preserve the original name, dependencies etc...
- The `phpbb/phpbb` requirement is automatically added (if it's not already present) and copied from `['extra']['soft-require']['phpbb/phpbb']`

Note1: It is waiting for a PR against the core to enable the ResonseListener. Otherwise the endpoints to get the json files doesn't work (didn't sent it yet because JIRA is broken)

### CODE REVIEW

**Comment by Nicofuma** [comment]
@VSEphpbb the issues should be solved now

**Comment by VSEphpbb** [comment]
Yep. Looks like package list is rebuilt on contrib approval :+1: 

It is doing some weird stuff with the Version check JSON data, ie: host: localhost

**Comment by Nicofuma** [comment]
that's intended, I think

File: includes/objects/revision.php [hunk 1]
```diff
@@ -706,19 +707,21 @@ public function update_composer_package($mode = 'add')
...
 
 		if ($mode == 'add')
 		{
-			$package_manager->add_release($this->revision_version, $this->attachment_id, true);
+			$package_manager->add_release(json_decode($this->revision_composer_json, true), $this->attachment_id, true);
 		}
 		else
 		{
-			$package_manager->remove_release($this->revision_version);
+			$package_manager->remove_release($this->revision_composer_json['version']);
```
Comment (prototech) [inline includes/objects/revision.php]: This is still a string and not an array. Last error here https://tracker.phpbb.com/browse/CUSTDB-697 refers to this.

File: includes/tools/composer_package_manager.php [hunk 2]
```diff
@@ -325,25 +334,21 @@ public function get_packages_group($write_map)
...
-			'version'	=> $version,
-			'dist'		=> 
-				array (
-					'url'	=> titania_url::build_clean_url('download', array('id' => $attachment_id)),
-					'type'	=> 'zip',
-			),
+		$controller_helper = phpbb::$container->get('phpbb.titania.controller.helper');
+
+		$composer_json['dist'] = array(
+			'url'	=> $controller_helper->route('phpbb.titania.download', array('id' => $attachment_id)),
```
Comment (prototech) [inline includes/tools/composer_package_manager.php]: There's a possibility of the user's sid being included in here. Use the path helper to strip that parameter.

File: controller/composer.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	* @return \Symfony\Component\HttpFoundation\Response
+	*/
+	public function serve_file($filename)
+	{
+		if (strpos($filename, '..') !== false)
+		{
+			throw new http_exception(404, 'NO_PAGE_FOUND');
+		}
+
+		$filename = \titania::$root_path . 'composer/' . $filename . '.json';
```
Comment (prototech) [inline controller/composer.php]: Use the `%phpbb.titania.root_path%` parameter instead of `\titania::$root_path`.

**Comment by prototech** [comment]
The version check data is based on the generated URL for the local installation. It's not hard-coded to point to phpBB.com.

File: includes/objects/revision.php (hunk 1 above)
Comment (Nicofuma) [inline includes/objects/revision.php]: fixed

File: includes/tools/composer_package_manager.php (hunk 2 above)
Comment (Nicofuma) [inline includes/tools/composer_package_manager.php]: fixed

File: controller/composer.php (hunk 3 above)
Comment (Nicofuma) [inline controller/composer.php]: fixed

**Comment by prototech** [comment]
@michaelcullum Please update the community board now that the required core PR has been merged.

**Comment by michaelcullum** [comment]
First of all it needs to be decided for sure that we are reversing the very concious decision in 3.1 to have a soft requirement differ from the hard composer requirement.

**Comment by Nicofuma** [comment]
@phpbb/development-team @VSEphpbb @Derky @paul999 @DavidIQ ^

**Comment by DavidIQ** [comment]
@michaelcullum can you elaborate on what you just said?

**Comment by nickvergessen** [comment]
We only moved to soft-require, because we didn't have phpbb set up on packagist back then?

**Comment by Derky** [comment]
Yeah that's what I thought as well. (and soft-require isn't part of the composer spec, it has just been made up for the reason Joas mentioned IIRC)

**Comment by DavidIQ** [comment]
Then there isn't really anything to hold this back, unless @michaelcullum has some additional information that would stop this from going any further.

**Comment by paul999** [comment]
@michaelcullum We are kinda waiting on this, can you please comment or can we otherwise merge it?

**Comment by Nicofuma** [comment]
By the way, @paul999, if someone want to use the packages.json file he needs to support the whole composer format because we may change it in the future (it could be needed if there if too much packages to serve). Also, this file may serve non phpBB package (a set of approved packagist packages per example).

In composer source code, the implementation related to the format we use is this one: https://github.com/composer/composer/blob/master/src/Composer/Repository/ComposerRepository.php (we don't use any provider for now, but it may change).

https://getcomposer.org/doc/05-repositories.md#composer

**Comment by paul999** [comment]
@Nicofuma I guess we can put that in the Titania documentation? :P

**Comment by Nicofuma** [comment]
That's a classic implementation following these "specs" https://getcomposer.org/doc/05-repositories.md#composer ;)

**Comment by paul999** [comment]
Ok. However we should still document somewhere in the titania documentation that we follow the standard specs?

**Comment by Nicofuma** [comment]
There is a documentation? =D I guess yes...

**Comment by michaelcullum** [comment]
Titania has no documentation really?

Sorry I haven't gotten around to responding to this yet, I'll respond to this PR tomorrow (I'm on my phone right now). This PR is not just waiting on me to comment on it as it also requires me to do some stuff to our phpbb installation too before it can be merged.

**Comment by paul999** [comment]
@michaelcullum this has some sort of priority as stuff (/users) is waiting on it.

**Comment by michaelcullum** [comment]
Sure, it's top of my phpBB to do list, but I've been busy with real life this last week, the exceptions being when websites keep dying and I've had to sort out stuff with OSL.

**Comment by Elsensee** [comment]
I'm not that much into composer but I got a question:
What happens if someone wants to install an extension which has phpbb as a requirement. Will the extension still be put in the correct place (like it's done when you set the extension as a requirement in your phpbb installation)?

**Comment by Nicofuma** [comment]
yep

**Comment by paul999** [comment]
@michaelcullum Bump

**Comment by DavidIQ** [comment]
This will be merged by 11 AM UCT-4 today unless there are any actual objections that don't include "I'll reply by x time/day".

**Comment by DavidIQ** [comment]
Command/feature to import already approved extensions to come later.

**Comment by michaelcullum** [comment]
Sure, because <3 hours is plenty of time. Merging this PR will also break phpBB if you actually read it, it requires some changes to phpBB, which also need me to make them. ;)

**Comment by DavidIQ** [comment]
@michaelcullum  < 3 hours???? You've had over _20 days_ to say what you needed to say and have not said anything useful whatsoever.

**Pull request merged**